README.rst
Torch Liberator - Deploy PyTorch Models 
---------------------------------------

|GitlabCIPipeline| |GitlabCICoverage| |Pypi| |Downloads| 

+----------------------+-------------------------------------------------------------+
| Main Page            | https://gitlab.kitware.com/computer-vision/torch_liberator  |
+----------------------+-------------------------------------------------------------+
| Github Mirror        | https://github.com/Kitware/torch_liberator                  |
+----------------------+-------------------------------------------------------------+
| Pypi                 | https://pypi.org/project/torch_liberator                    |
+----------------------+-------------------------------------------------------------+
| Torch Hackathon 2021 | `Youtube Video`_, `Google Slides`_, and `Submission Page`_  |
+----------------------+-------------------------------------------------------------+

.. _Youtube Video: https://www.youtube.com/watch?v=GQqtn61iNsc
.. _Google Slides: https://docs.google.com/presentation/d/1w9XHkPjtLRj29dw50WP0rSHRRlEfhksP_Sf8XldTSYE
.. _Submission Page: https://devpost.com/software/torchliberator-partial-weight-loading

Torch Liberator is a Python module containing a set of tools for reading and
writing relevant parts of deep networks.

Typically, when moving a deep-network trained with torch, you have to keep
track of the entire codebase that defined the model file in addition to the
checkpoint file containing the learned weights. Torch Liberator contains tools
to extract relevant source code and bundle it with weights and serializing it
into a single-file deployment.

Note: as of torch 1.9, torch comes with a `torch.package <https://pytorch.org/docs/stable/package.html>`__ 
submodule which contains a method for saving model weights with model
structure.  We recommend using ``torch.package`` over the single-file deployments
provided in this package. Thus the ``load_partial_state`` logic is the main
code of interest provided in this module.



Installation
------------

.. code:: bash

    pip install torch_liberator

    # OR with a specific branch

    pip install git+https://gitlab.kitware.com/computer-vision/torch_liberator.git@main


Partial State Loading
---------------------

New in ``0.1.0`` torch liberator now exposes a public ``load_partial_state``
function, which does it best to "shove" weights from one model into another
model. There are several methods to compute associations between layer names in
one model to layer names in another, the most general being the "embedding"
method, and the slightly more structured "isomorphism" option.


Have you ever had the scenario where you use one model as a sub-model in a
bigger network? Then you had to load pretrained subnetwork state into that
bigger model? 

The latest version of ``torch_liberator.load_patial_state`` can handle this by
solving a maximum-common-subtree-isomorphism problem. This computes the largest
possible mapping between the two state dictionaries that share consistent
suffixes.

.. code:: python 

    >>> import torchvision
    >>> import torch
    >>> import torch_liberator
    >>> resnet50 = torchvision.models.resnet50()
    >>> class CustomModel(torch.nn.Module):
    >>>     def __init__(self):
    >>>         super().__init__()
    >>>         self.module = resnet50
    >>>         self.extra = torch.nn.Linear(1, 1)
    >>> # Directly load resnet50 state into a model that has it as an embedded subnetwork
    >>> model = CustomModel()
    >>> model_state_dict = resnet50.state_dict()
    >>> # load partial state returns information about what it did
    >>> info = torch_liberator.load_partial_state(model, model_state_dict, association='isomorphism', verbose=1)
    >>> print(len(info['seen']['full_add']))
    >>> print(len(info['self_unset']))
    >>> print(len(info['other_unused']))
    320
    2
    0
    

It can also handle loading common state between two models that share some
underlying structure.

.. code:: python 

    >>> import torchvision
    >>> import torch
    >>> import torch_liberator
    >>> resnet50 = torchvision.models.resnet50()
    >>> class CustomModel1(torch.nn.Module):
    >>>     def __init__(self):
    >>>         super().__init__()
    >>>         self.module = resnet50
    >>>         self.custom_model1_layer = torch.nn.Linear(1, 1)
    >>> class CustomModel2(torch.nn.Module):
    >>>     def __init__(self):
    >>>         super().__init__()
    >>>         self.backbone = resnet50
    >>>         self.custom_model2_layer = torch.nn.Linear(1, 1)
    >>> # Load as much of model1 state into model2 as possible
    >>> model1 = CustomModel1()
    >>> model2 = CustomModel2()
    >>> model2_state_dict = model2.state_dict()
    >>> # load partial state returns information about what it did
    >>> info = torch_liberator.load_partial_state(model1, model2_state_dict, association='isomorphism', verbose=1)
    >>> print(len(info['seen']['full_add']))
    >>> print(len(info['seen']['skipped']))
    >>> print(len(info['self_unset']))
    >>> print(len(info['other_unused']))
    320
    2
    2
    2


.. code:: python 


    >>> import torchvision
    >>> import torch_liberator
    >>> #
    >>> faster_rcnn = torchvision.models.detection.faster_rcnn.fasterrcnn_resnet50_fpn()
    >>> resnet50 = torchvision.models.resnet50(pretrained=True)
    >>> state_dict = resnet50.state_dict()
    >>> # Load partial state return a dictionary that tells you how well it did
    >>> info = torch_liberator.load_partial_state(faster_rcnn, state_dict, verbose=0, association='embedding')
    >>> print(ub.map_vals(len, info['seen']))
    >>> print(ub.map_vals(len, ub.dict_diff(info, ['seen'])))
    {'full_add': 265, 'skipped': 55}
    {'other_unused': 55, 'self_unset': 30}

    >>> # Load partial state return a dictionary that tells you how well it did
    >>> info = torch_liberator.load_partial_state(faster_rcnn, state_dict, verbose=0, association='isomorphism')
    >>> print(ub.map_vals(len, info['seen']))
    >>> print(ub.map_vals(len, ub.dict_diff(info, ['seen'])))
    {'full_add': 265, 'skipped': 55}
    {'other_unused': 55, 'self_unset': 30}
    
    

Also, if the sizes of the tensor don't quite fit, they will be mangled, i.e.
"shoved-in" as best as possible. See the docstring for more detail.


Stand-alone Single-File Model Deployments
-----------------------------------------

The original purpose of ``torch_liberator`` was to build standalone torch
packages that contained both the model code and the model weight. It still does
that but ``torch.package`` new in torch 1.9, might be a better solution moving
forward. See `torch.package <https://pytorch.org/docs/stable/package.html>`__
for details.

Torch Liberator builds on the
`liberator <https://gitlab.kitware.com/python/liberator>`__ library to statically
extract pytorch code that defines a model's topology and bundle that with a
pretrained weights file. This results in a single-file deployment package and
can potentially remove dependencies on the codebase used to train the model.

Torch Liberator can also read these deployment files and create an instance of
the model initialized with the correct pretrained weights.

The API is ok, but it does need improvement. However, the current version is in
a working state. There aren't any high level docs, but there are a lot of
docstrings and doctests. The example here gives a good overview of the code by
extracting the AlexNet model from torchvision.


.. code:: python 

    >>> import torch_liberator
    >>> from torch_liberator.deployer import DeployedModel
    >>> from torchvision import models

    >>> print('--- DEFINE A MODEL ---')
    >>> model = models.alexnet(pretrained=False)  # false for test speed
    >>> initkw = dict(num_classes=1000)  # not all models nicely supply this
    >>> model._initkw = initkw
    --- DEFINE A MODEL ---

    >>> print('--- DEPLOY THE MODEL ---')
    >>> zip_fpath = torch_liberator.deploy(model, 'test-deploy.zip')
    --- DEPLOY THE MODEL ---
    [DEPLOYER] Deployed zipfpath=/tmp/tmpeqd3y_rx/test-deploy.zip
    

    >>> print('--- LOAD THE DEPLOYED MODEL ---')
    >>> loader = DeployedModel(zip_fpath)
    >>> model = loader.load_model()
    --- LOAD THE DEPLOYED MODEL ---
    Loading data onto None from <zopen(<_io.BufferedReader name='/tmp/tmpg1kln3kw/test-deploy/deploy_snapshot.pt'> mode=rb)>
    Pretrained weights are a perfect fit
    

The major weirdness right now, is you either have to explicitly define "initkw"
(which are the keyword arguments used to create an instance of our model) at
deploy time, or you can set it as the ``_initkw`` attribute of your model (or
if your keyword arguments all exist as member variables of the class,
torch_liberator tries to be smart and infer what initkw should be).


There is also a torch-liberator CLI that can be used to package a weight file,
a python model file, and optional json metadata.

.. code:: bash

    python -m torch_liberator \
        --model <path-to-the-liberated-py-file> \
        --weights <path-to-the-torch-pth-weight-file> \
        --info <path-to-train-info-json-file> \
        --dst my_custom_deployfile.zip


.. |Pypi| image:: https://img.shields.io/pypi/v/torch_liberator.svg
   :target: https://pypi.python.org/pypi/torch_liberator

.. |Downloads| image:: https://img.shields.io/pypi/dm/torch_liberator.svg
   :target: https://pypistats.org/packages/torch_liberator

.. |ReadTheDocs| image:: https://readthedocs.org/projects/torch_liberator/badge/?version=latest
    :target: http://torch_liberator.readthedocs.io/en/latest/

.. # See: https://ci.appveyor.com/project/jon.crall/torch_liberator/settings/badges
.. .. |Appveyor| image:: https://ci.appveyor.com/api/projects/status/py3s2d6tyfjc8lm3/branch/main?svg=true
.. :target: https://ci.appveyor.com/project/jon.crall/torch_liberator/branch/main

.. |GitlabCIPipeline| image:: https://gitlab.kitware.com/computer-vision/torch_liberator/badges/main/pipeline.svg
   :target: https://gitlab.kitware.com/computer-vision/torch_liberator/-/jobs

.. |GitlabCICoverage| image:: https://gitlab.kitware.com/computer-vision/torch_liberator/badges/main/coverage.svg?job=coverage
    :target: https://gitlab.kitware.com/computer-vision/torch_liberator/commits/main



CHANGELOG.md
# Changelog

We are currently working on porting this changelog to the specifications in
[Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## Version 0.2.2 - Unreleased

### Changed
* Minor internal details


## Version 0.2.1 - Released 2023-02-11

### Fixed
* Work around networkx removal of OrderedDiGraph


## Version 0.2.0 - Released 2022-09-16

### Changed:

* Moved implementation of the backend maximum common subtree algorithm to `https://github.com/Erotemic/networkx_algo_common_subtree`.
* The weight mapping is now returned by the partial state loading and pretrained forward
* Changed defaults to use the isomorphism algorithm by default and other tweaks.


## Version 0.1.1 - Released 2021-11-01

### Changed

* Using scikit-build and CMake to build cython modules and publishing wheels
  instead of hacky autojit solutions.


## Version 0.1.0 - Released 2021-05-06

### Added
* Ported the maximum common subtree isomorphism building associations in `load_partial_state`.


## Version 0.0.5 - Released 2021-05-06

### Added
* Added `cli` for manually making deployed zip files

### Changed
* Removed Python 2.7 and 3.5 Support.
* The first part in the deployed name is now based on the netharn run "name"
  rather than the name of the model class

### Fixed
* Support new netharn checkpoints dir


## Version 0.0.4 - Released 2021-05-06

### Added
* Added `extract_snapshot` method to `DeployedModel`


## Version 0.0.3 - Released


## [Version 0.0.1] - 

### Added
* Initial version



dev/example_partial.py
# flake8: noqa


def example1():
    """
    import torchvision
    import torch_liberator
    # Create two similar but different models
    faster_rcnn = torchvision.models.detection.fasterrcnn_resnet50_fpn()
    model = torchvision.models.resnet50(pretrained=True)

    faster_rcnn.load_state_dict(model.state_dict())

    # Return a dictionary that tells what load_partial_state did
    info = torch_liberator.load_partial_state(
         faster_rcnn, model.state_dict(),
         association='isomorphism')

    >>> print(ub.map_vals(len, info['seen']))
    >>> print(ub.map_vals(len, ub.dict_diff(info, ['seen'])))
    {'full_add': 265, 'skipped': 55}
    {'other_unused': 55, 'self_unset': 30}
    """



































def example2():
    """
    >>> # Load partial state return a dictionary that tells you how well it did
    >>> info = torch_liberator.load_partial_state(
    >>>     faster_rcnn, state_dict, verbose=0, association='isomorphism')
    >>> print(ub.map_vals(len, info['seen']))
    >>> print(ub.map_vals(len, ub.dict_diff(info, ['seen'])))
    {'full_add': 265, 'skipped': 55}
    {'other_unused': 55, 'self_unset': 30}


    """


def has_this_ever_happened_to_you():

    import torchvision
    import torch
    raw_model = torchvision.models.resnet18(pretrained=True)

    dp_model = torch.nn.DataParallel(raw_model, device_ids=[0])
    dp_model.load_state_dict(raw_model.state_dict())

    """
    RuntimeError: Error(s) in loading state_dict for DataParallel:
    Missing key(s) in state_dict: "module.conv1.weight",
    """

    # FIX IT
    import torch_liberator
    _ = torch_liberator.load_partial_state(dp_model, raw_model.state_dict(), association='isomorphism')
    """
    Pretrained weights are a perfect fit
    """

    rank = 0
    world_size = 1
    import os
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12355'
    torch.distributed.init_process_group("gloo", rank=rank, world_size=world_size)
    ddp_model = torch.nn.parallel.distributed.DistributedDataParallel(raw_model, device_ids=[0])
    ddp_model.load_state_dict(raw_model.state_dict())

    """
    RuntimeError: Error(s) in loading state_dict for DistributedDataParallel:
    Missing key(s) in state_dict: "module.conv1.weight", "module.bn1.weight"
    """


def visualization():
    import torchvision
    import torch_liberator
    import networkx as nx
    import kwplot
    import kwimage
    import graphid
    import torch
    from torchvision.models.detection.backbone_utils import resnet_fpn_backbone
    kwplot.autoplt()

    clf_model = torchvision.models.resnet18(pretrained=True)

    clf_model2 = torchvision.models.resnet18(pretrained=False)
    clf_model2.conv1 = torch.nn.Conv2d(13, 64, 7, 2, 3)
    clf_model2.fc = torch.nn.Linear(512, 20)

    det_model = resnet_fpn_backbone('resnet18', pretrained=False)

    wrp_model = torch.nn.DataParallel(clf_model, device_ids=[0])

    def model_to_simplified_paths(model, name='.'):
        paths = [name + '.' + p for p in model.state_dict().keys()]
        paths = [p for p in paths if not p.endswith('num_batches_tracked')]
        paths = [p for p in paths if not p.endswith('running_var')]
        paths = [p for p in paths if not p.endswith('running_mean')]
        paths = [p for p in paths if not p.endswith('bias')]
        paths = [p[:-len('.weight')] if p.endswith('.weight') else p for p in paths]
        # paths = [p.replace('.weight', '') for p in paths]
        paths = [p for p in paths if 'layer2' not in p]
        paths = [p for p in paths if 'layer3' not in p]
        return paths

    clf_paths = model_to_simplified_paths(clf_model, 'clf')
    det_paths = model_to_simplified_paths(det_model, 'det')

    clf_tree = torch_liberator.initializer.paths_to_otree(clf_paths, sep='.')
    det_tree = torch_liberator.initializer.paths_to_otree(det_paths, sep='.')

    clf_tree = torch_liberator.initializer.paths_to_otree(list(clf_model.state_dict().keys()), sep='.')
    det_tree = torch_liberator.initializer.paths_to_otree(list(det_model.state_dict().keys()), sep='.')

    from torch_liberator import _nx_ext_v2
    node_affinity = torch_liberator.initializer._common_suffix_affinity
    subtree1, subtree2, value = _nx_ext_v2.maximum_common_ordered_subtree_isomorphism(
        clf_tree, det_tree, node_affinity=node_affinity)

    wrp_paths = model_to_simplified_paths(wrp_model, 'wrp')
    wrp_tree = torch_liberator.initializer.paths_to_otree(wrp_paths, sep='.')

    nx.set_node_attributes(wrp_tree, name='color', values=kwimage.Color('lightpink').as01())
    nx.set_node_attributes(clf_tree, name='color', values=kwimage.Color('lightblue').as01())
    nx.set_node_attributes(det_tree, name='color', values=kwimage.Color('lightgreen').as01())

    layoutkw = {'rankdir': 'LR'}
    kwplot.figure(fnum=1, doclf=True)
    _ = graphid.util.util_graphviz.show_nx(
        clf_tree,
        fnum=1,
        as_directed=False,
        fontsize=10,
        layoutkw=layoutkw,
    )

    kwplot.figure(fnum=2, doclf=True)
    _ = graphid.util.util_graphviz.show_nx(
        det_tree,
        fnum=2,
        as_directed=False,
        fontsize=10,
        layoutkw=layoutkw,
    )

    kwplot.figure(fnum=4, doclf=True)
    _ = graphid.util.util_graphviz.show_nx(
        wrp_tree,
        fnum=4,
        as_directed=False,
        fontsize=10,
        layoutkw=layoutkw,
    )

    clf_tree = torch_liberator.initializer.paths_to_otree(clf_paths, sep='/')
    det_tree = torch_liberator.initializer.paths_to_otree(det_paths, sep='/')
    nx.set_node_attributes(clf_tree, name='color', values=kwimage.Color('lightblue').as01())
    nx.set_node_attributes(det_tree, name='color', values=kwimage.Color('lightgreen').as01())
    _ = graphid.util.util_graphviz.nx_agraph_layout(clf_tree, inplace=1, **layoutkw)  # NOQA
    _ = graphid.util.util_graphviz.nx_agraph_layout(det_tree, inplace=1, **layoutkw)  # NOQA
    stacked = graphid.util.util_graphviz.stack_graphs([clf_tree, det_tree], vert=False)
    for n1, n2 in zip(subtree1.nodes, subtree2.nodes):
        stacked.add_edge(n1, n2, implicit=True, color=kwimage.Color('orange').as01())
    # stacked.add_edge(('clf',), ('det',), implicit=True)

    # Does only neato respect pin?
    layoutkw_ = layoutkw.copy()
    layoutkw_['prog'] = 'neato'
    _ = graphid.util.util_graphviz.nx_agraph_layout(stacked, inplace=True, **layoutkw_)  # NOQA

    layout_info = graphid.util.util_graphviz.get_nx_layout(stacked, 'custom')
    kwplot.figure(fnum=5).gca().cla()
    _ = graphid.util.util_graphviz.show_nx(
        stacked,
        as_directed=False,
        fnum=3,
        layout='custom',
        fontsize=10,
        layoutkw=layoutkw_, verbose=4)

    # if 0:
    #     agraph = graphid.util.util_graphviz.make_agraph(stacked2.copy())
    #     layoutkw_ = layoutkw.copy()
    #     prog = layoutkw_.pop('prog', 'dot')
    #     argparts = ['-G%s=%s' % (key, str(val))
    #                 for key, val in layoutkw.items()]
    #     args = ' '.join(argparts)
    #     agraph.layout(prog=prog, args=args)
    #     agraph.draw('foo.png')
    #     import xdev
    #     xdev.startfile('foo.png')
